raft_consensus/types.nim

```diff
@@ -0,0 +1,121 @@
+                                                # providing at minimum operations for initialization, querying the current state
+                                                # and RAFTNodeLogEntry application
+        state: RAFTNodeStateMachineState
+
+    # RAFT Node Persistent Storage basic definition
+    RAFTNodePersistentStorage* = ref object     # Should be some kind of Persistent Transactional Store Wrapper
+
+    # Basic modules (algos) definitions
+    RAFTNodeAccessCallback = proc: RAFTNode {.nimcall, gcsafe.}     # This should be implementes as a closure holding the RAFTNode
+    RAFTConsensusModule* = object of RootObj
+        state_transitions_fsm: seq[byte]        # I plan to use nim.fsm https://github.com/ba0f3/fsm.nim
+        raft_node_access_callback: RAFTNodeAccessCallback
+
+    RAFTLogCompactionModule* = object of RootObj
+        raft_node_access_callback: RAFTNodeAccessCallback
+
+    RAFTMembershipChangeModule* = object of RootObj
+        raft_node_access_callback: RAFTNodeAccessCallback
+
+    # Callback for sending messages out of this RAFT Node
+    RAFTMessageId* = UUID                       # UUID assigned to every RAFT Node Message,
+                                                # so it can be matched with it's coresponding response etc.
+
+    RAFTMessageSendCallback* = proc (raft_message: RAFTMessageBase) {.nimcall, gcsafe.} # Callback for Sending RAFT Node Messages
+                                                                                        # out of this RAFT Node. Can be used for broadcasting
+                                                                                        # (a Heart-Beat for example)
+
+    # RAFT Node basic Log definitions
+    RAFTNodeLogEntry* = ref object              # Abstarct RAFT Node Log entry containing opaque binary data (Blob)
+        term*: RAFTNodeTerm
+        data*: Blob
+
+    RAFTNodeLog* = ref object                   # Needs more elaborate definition. Probably this will be a RocksDB/MDBX/SQLite Store Wrapper etc.
+        log_data*: seq[RAFTNodeLogEntry]        # RAFT Node Log Data
+
+    # Base typoe for RAFT message objects
+    RAFTMessageBase* = ref object of RootObj    # Base Type for RAFT Node Messages
+        msg_id*: RAFTMessageId                  # Message UUID
+        sender_id*: RAFTNodeId                  # Sender RAFT Node ID
+        sender_term*: RAFTNodeTerm              # Sender RAFT Node Term
+        peers*: RAFTNodePeers                   # List of RAFT Node IDs, which should receive this message
+
+    # RAFT Node Object definitions
+    RAFTNode* = object
+        # Timers
+        voting_timout: uint64
+        heart_beat_timeout: uint64
+        # etc. timers
+
+        # Mtx definitions go here
+        raft_state_mutex: Lock
+        raft_log_mutex: Lock
+        raft_comm_mutex_receive_msg: Lock
```

Comment:
Let's follow the [NEP1 naming conventions](https://nim-lang.org/1.0.0/nep1.html#introduction-naming-conventions) (i.e. use `camelCase` for field names).

Reply:
Got it.